### ISSUE TITLE
Beets-like sorting and naming?

### ISSUE THREAD

**Issue by naelstrof** [contrixed/RomReader#1]
This is actually a really neat idea, and it reminds me of another program I really like called [beets](http://beets.radbox.org/).
Roms are nearly as plentiful and as varied as music, so what do you think about enforcing a naming convention, or managing a separate "rom library", like beets, with proper names and neat automatic organization (games sorted into folders by system, director, and dev team or something).

Having support for ProjectM roms and being able to validate them would save a lot of headaches for some people as well.

This is a really cool idea, I hope it goes somewhere. Keep up the good work!

**Comment by contrixed** [comment]
There is quite a bit that can be done with this concept; this project is actually a piece of another project that is currently being planned out. I was considering a database considering how identifying Playstation and Playstation 2 Rom format has its naming scheme for example `SLUS21315 = 50cent - Bulletproof`.  This is an interesting idea however, feel free to contribute in anyway, this project is very much alive.

**Comment by naelstrof** [comment]
Yeah I'd love to contribute. Since RomReader is part of the bigger whole, I would need some kind of design document that describes the planned features and structure of the bigger project. It's your project so I don't want to make bad decisions for you.

I think it wouldn't be too hard to create a database of checksums, and to organize recognized roms into a central library.

Problem is thinking up a standard naming scheme. The different release versions of games really muck things up...

**Comment by contrixed** [comment]
The second project is only going to use RomReader as a library, RomReader is not going to migrate into it. The database i have an idea for in terms of structure, As for the naming scheme follow goodset naming, after all, most roms are named in the goodset naming scheme, except for archives and loose roms that you find here and there. 

The other project is similar to GoodTools except its only target platform is Linux and other Unix variants. The funny thing is, the python version is the Prototype, The C/C++ version is where all the most interesting features will go. The Python variant will be maintained after the C variant is out, as both versions will be different; E.G Python versions will only be a cli app as the C/C++ version will have a gui and many other features that have been decided on the C/C++ version. 

But mind you the C/C++ version is far far away, the only focus is to get the Python variant completed then we have our concept and logic for the C/C++ variant. 

I welcome most changes, this is far from a beta, even if the current structure changes that's fine. (The structure already changed, just have to finish a few things before i push) If you would like me to, i can add you as a collaborator in the repository. This way we cut the need out for PRs.

feel free to ask anything that you may have concerns or questions about. 

Thanks for being interested!

**Comment by naelstrof** [comment]
Sounds awesome, a few things to mention though:

> E.G Python versions will only be a cli app as the C/C++ version will have a gui and many other features that have been decided on the C/C++ version. 

It'd be a good idea to keep the program a completely CLI one, and to add a GUI wrapper for it later which may be what you have planned, but it doesn't seem very clear, I'm guessing you have some feature planned that can't be replicated in CLI, regardless keeping CLI separate from GUI would separate dependencies, and headaches.

> The C/C++ version is where all the most interesting features will go.

Are you sure python isn't good enough for everything? C++ string/file/memory management sucks, and as a bonus python is pretty much automatically multi-platform. I do accel at C++ programming though so I don't really care.

> The other project is similar to GoodTools except its only target platform is Linux and other Unix variants.

You should really plan for Windows/Mac support, basic file operations are already multi-platform, and as long as you stick to some standard GUI stuff like GTK or QT you won't have to do very much porting, if at all.

I can help you set up a build system that makes it pretty easy to keep your stuff multi-platform, but it depends on which language you're going to end up using.

> If you would like me to, i can add you as a collaborator in the repository. This way we cut the need out for PRs.

Don't, I still don't know exactly what you have planned. You're talking of big things that seem to be outside the scope of this repository. It's raising quite a few questions for me in order to help you, but let me get a few things straight:
* I'm only interested in the ROM identification, naming, and library management. Similar to how beets manages music.
* Where could I help start, or otherwise implement this functionality?
* How does my planned features interact with yours? Would mine be a separate project, a repository that I made, that uses RomReader to help identify/verify roms? Or are you going to make another repository in which these features can exist?

**Comment by contrixed** [comment]
> It'd be a good idea to keep the program a completely CLI one, and to add a GUI wrapper for it later which may be what you have planned, but it doesn't seem very clear, I'm guessing you have some feature planned that can't be replicated in CLI, regardless keeping CLI separate from GUI would separate dependencies, and headaches.

RomReader will not have a gui, i should have been more in depth, the main project will contain a gui but not the Python variant.

> Are you sure python isn't good enough for everything? C++ string/file/memory management sucks, and as a bonus python is pretty much automatically multi-platform. I do accel at C++ programming though so I don't really care.

The project is intended to be ran on multiple device types, right now the first device is a Raspberry Pi, the Python version will be able to get some of the features in a later date.

> You should really plan for Windows/Mac support, basic file operations are already multi-platform, and as long as you stick to some standard GUI stuff like GTK or QT you won't have to do very much porting, if at all.

At a later time, and if you would like to help with that so later on why not. The main reason i say  Linux/Unix is the main target atm is me and the other person behind just run Linux mainly. I myself only run Linux.

> Don't, I still don't know exactly what you have planned. You're talking of big things that seem to be outside the scope of this repository. It's raising quite a few questions for me in order to help you, but let me get a few things straight:

Fair enough. Let me explain here;

> * Where could I help start, or otherwise implement this functionality?

You can help by making more readers, for example we are missing quite a few, we have quite a few readers, however due to the support i intend to support i am working on those features They include:
* Archive support for goodsets
* Simple IO wrapper that will automate archives
* Determining a default file based on preferences of goodsets.

The beets functionality can be done one a separate, we could have an organization and split RomReader and <unknown_project_name> and incorporate them. 

I have seen the beets source code, and understand it to a decent point of what i saw, so i would have no problem helping out at all. 

Or we could include the functionality with RomReader as a package for it, but there is some consideration with this option. In any case i welcome these two, or if you have an idea share it, i have no problems with advice or opinions.

**Comment by naelstrof** [comment]
Ahh I understand perfectly now. Thanks! I'll fork the project and mess around for a bit. Don't expect anything fast.

**Comment by contrixed** [comment]
Take your time, shortly ill have the archive support pushed, just have to finish writing the unittests.
